src/server/v2.0/handler/project.go

```diff
@@ -63,3 +241,380 @@ func (a *projectAPI) GetLogs(ctx context.Context, params operation.GetLogsParams
+		fetchSummaries = append(fetchSummaries, getProjectQuotaSummary)
+	}
+
+	if hasPerm := a.HasProjectPermission(ctx, p.ProjectID, rbac.ActionList, rbac.ResourceMember); hasPerm {
+		fetchSummaries = append(fetchSummaries, getProjectMemberSummary)
+	}
+
+	if p.RegistryID > 0 {
+		fetchSummaries = append(fetchSummaries, getProjectRegistrySummary)
+	}
+
+	var wg sync.WaitGroup
+	for _, fn := range fetchSummaries {
+		fn := fn
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			fn(ctx, p, summary)
+		}()
+	}
+	wg.Wait()
+
+	return operation.NewGetProjectSummaryOK().WithPayload(summary)
+}
+
+func (a *projectAPI) HeadProject(ctx context.Context, params operation.HeadProjectParams) middleware.Responder {
+	if err := a.RequireAuthenticated(ctx); err != nil {
+		return a.SendError(ctx, err)
+	}
+
+	if _, err := a.projectCtl.GetByName(ctx, params.ProjectName); err != nil {
+		return a.SendError(ctx, err)
+	}
+
+	return operation.NewHeadProjectOK()
+}
+
+func (a *projectAPI) ListProjects(ctx context.Context, params operation.ListProjectsParams) middleware.Responder {
+	query := q.New(q.KeyWords{})
+	query.Sorting = "name"
+	query.PageNumber = *params.Page
+	query.PageSize = *params.PageSize
+
+	if name := lib.StringValue(params.Name); name != "" {
+		query.Keywords["name"] = &q.FuzzyMatchValue{Value: name}
+	}
+	if owner := lib.StringValue(params.Owner); owner != "" {
+		query.Keywords["owner"] = owner
+	}
+	if params.Public != nil {
+		query.Keywords["public"] = lib.BoolValue(params.Public)
+	}
+
+	secCtx, ok := security.FromContext(ctx)
+	if ok && secCtx.IsAuthenticated() {
+		if !secCtx.IsSysAdmin() && !secCtx.IsSolutionUser() {
+			// authenticated but not system admin or solution user,
+			// return public projects and projects that the user is member of
+			if l, ok := secCtx.(*local.SecurityContext); ok {
+				currentUser := l.User()
+				member := &project.MemberQuery{
+					Name:     currentUser.Username,
+					GroupIDs: currentUser.GroupIDs,
+				}
+
+				// not filter by public or filter by the public with true,
+				// so also return public projects for the member
+				if public, ok := query.Keywords["public"]; !ok || lib.ToBool(public) {
+					member.WithPublic = true
+				}
+
+				query.Keywords["member"] = member
+			} else {
+				// can't get the user info, force to return public projects
+				query.Keywords["public"] = true
+			}
+		}
+	} else {
+		if params.Public != nil && !*params.Public {
+			// anonymous want to query private projects return empty projects directly
+			return operation.NewListProjectsOK().WithXTotalCount(0).WithPayload([]*models.Project{})
+		}
+		// force to return public projects for anonymous
+		query.Keywords["public"] = true
+	}
+
+	total, err := a.projectCtl.Count(ctx, query)
+	if err != nil {
+		return a.SendError(ctx, err)
+	}
+
+	projects, err := a.projectCtl.List(ctx, query, project.WithCVEAllowlist(), project.WithOwner())
```

Comment:
why not use len(projects) as total to reduce one DB call.

Reply:
This API support pagination, the total is the `total` count of projects for the request, `projects` is only the current page data.

En, it's a good point that when the total is 0, it does not need to call the `List` to get the current page data.